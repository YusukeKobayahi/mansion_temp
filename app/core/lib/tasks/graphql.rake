# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
