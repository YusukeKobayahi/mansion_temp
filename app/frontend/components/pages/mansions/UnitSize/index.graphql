fragment MsLibraryMansionsUnitSize on Mansion {
  name
  salesHistorySummaries {
    layout
    numberOfRooms
    layoutKind {
      slug
      sort
    }
    exclusiveAreas
  }
}
