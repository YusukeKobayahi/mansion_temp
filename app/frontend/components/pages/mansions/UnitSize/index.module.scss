@import "~/styles/utils";

.container {
  margin-top: 50px;
  padding: 0 40px;

  @include screenMd {
    padding: 0 16px;
  }
}

.info {
  line-height: 1.5;
}

.table {
  width: 100%;
  border-bottom: 1px solid #d0d0d0;
  margin-top: 32px;

  td {
    padding: 15px;
    border-top: 1px solid #d0d0d0;
    line-height: 1.5;
    background-clip: padding-box;
  }

  th {
    width: 144px;
    padding: 15px;
    border-top: 1px solid #d0d0d0;
    background-color: #faf9f7;
    line-height: 1.5;
    text-align: center;
    vertical-align: middle;

    // firefoxで枠線が表示されない問題の対応
    // See: https://qiita.com/h-koeda/items/3b4c6a465ff4f81c16dc
    background-clip: padding-box;

    &.hasUnit {
      position: relative;
      background-color: #edf1f4;

      &::before {
        content: "部屋有";
        display: block;
        position: absolute;
        left: 6px;
        top: 50%;
        transform: translateY(-50%); // 縦方向に中央表示
        width: 22px;
        padding: 3px 0;
        border: 1px solid #707070;
        background-color: #ffffff;
        font-size: 12px;
        line-height: 1;
        text-align: center;
      }
    }
  }

  @include screenMd {
    td {
      display: block;
      width: 100%;
      padding: 11px 9px;
    }

    th {
      display: block;
      width: 100%;
      padding: 6px;

      &.hasUnit {
        &::before {
          left: auto;
          right: 7px;
          width: auto;
          padding: 3px;
        }
      }
    }
  }
}
