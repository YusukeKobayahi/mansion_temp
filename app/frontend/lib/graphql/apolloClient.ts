// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
