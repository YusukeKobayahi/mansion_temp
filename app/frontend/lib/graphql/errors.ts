export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
