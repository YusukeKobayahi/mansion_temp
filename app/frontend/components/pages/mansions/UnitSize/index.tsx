import { useMemo } from "react";

import { MsLibraryMansionsUnitSizeFragment } from "~/graphql/generated";
import SidelineHeading from "~/components/commons/headings/SidelineHeading";
import styles from "./index.module.scss";
import uniq from "lodash/uniq";

export interface UnitSizeProps {
  data: MsLibraryMansionsUnitSizeFragment;
}

function createSuitableUseText(
  numberOfRooms: number,
  layoutKindSlug: string
): string {
  switch (numberOfRooms) {
    case 1:
      if (layoutKindSlug === "r" || layoutKindSlug === "k") {
        return "一人暮らしを検討している方";
      }
      return "部屋を活用する一人暮らし・二人暮らしを検討している方";
    case 2:
      return "ゆったりとした二人暮らしを検討している方や、お子様がお一人いるご家庭";
    case 3:
      return "お子様が二人いるご家庭で広々と部屋をご活用したい場合";
    default:
      return "お子様が二人以上いて子供部屋が欲しいとお考えの場合や住むご予定の人数が多い場合";
  }
}

const UnitSize: React.FC<UnitSizeProps> = ({ data }: UnitSizeProps) => {
  const content = useMemo(() => {
    // 間取りの小さい順に並び替えた MansionSalesHistorySummary
    const sortedSummaries = data.salesHistorySummaries.sort((a, b) => {
      if (a.numberOfRooms !== b.numberOfRooms) {
        return a.numberOfRooms - b.numberOfRooms;
      }
      return a.layoutKind.sort - b.layoutKind.sort;
    });

    // 小さい順に並び替えた全 MansionSalesHistorySummary の専有面積
    const sortedAreas = data.salesHistorySummaries
      .reduce<number[]>((areas, val) => areas.concat(val.exclusiveAreas), [])
      .sort((a, b) => a - b);

    const minLayoutSummary = sortedSummaries[0];
    const maxLayoutSummary = sortedSummaries[sortedSummaries.length - 1];
    const minExclusiveArea = sortedAreas[0];
    const maxExclusiveArea = sortedAreas[sortedAreas.length - 1];

    const layoutRangeText = uniq([
      minLayoutSummary.layout,
      maxLayoutSummary.layout,
    ]).join("～");

    const suitableUseRangeText = uniq([
      createSuitableUseText(
        minLayoutSummary.numberOfRooms,
        minLayoutSummary.layoutKind.slug
      ),
      createSuitableUseText(
        maxLayoutSummary.numberOfRooms,
        maxLayoutSummary.layoutKind.slug
      ),
    ]).join("～");

    const exclusiveAreaRangeText = uniq([
      `${minExclusiveArea}㎡`,
      `${maxExclusiveArea}㎡`,
    ]).join("～");

    const hasUnitOf39OrLess = sortedAreas.some((area) => area < 40);
    const hasUnitOf40To59 = sortedAreas.some((area) => 40 <= area && area < 60);
    const hasUnitOf60To89 = sortedAreas.some((area) => 60 <= area && area < 90);
    const hasUnitOf90OrMore = sortedAreas.some((area) => 90 <= area);

    return (
      <>
        <p className={styles.info}>
          間取りには{layoutRangeText}があり、{suitableUseRangeText}
          におすすめのマンションです。
          <br />
          広さでみると{exclusiveAreaRangeText}
          の物件があり、{data.name}
          では表内の色がついている広さの部屋が存在しています。
        </p>
        <table className={styles.table}>
          <tbody>
            <tr>
              <th className={hasUnitOf39OrLess ? styles.hasUnit : ""}>
                ～39㎡
              </th>
              <td>
                一人暮らしに向いており、特に30㎡以上だとゆったり使えるスペースがあります。
              </td>
            </tr>
            <tr>
              <th className={hasUnitOf40To59 ? styles.hasUnit : ""}>
                40～59㎡
              </th>
              <td>
                二人暮らしに向いている広さです。リビング・ダイニングの他に1,2部屋つく目安です。
              </td>
            </tr>
            <tr>
              <th className={hasUnitOf60To89 ? styles.hasUnit : ""}>
                60～89㎡
              </th>
              <td>
                家族3人でも過ごしやすく、お子様がいる場合にもおすすめです。
              </td>
            </tr>
            <tr>
              <th className={hasUnitOf90OrMore ? styles.hasUnit : ""}>
                90㎡～
              </th>
              <td>
                家族4人以上で暮らす場合にも適しており、色々な用途でスペースを使い分けられます。
              </td>
            </tr>
          </tbody>
        </table>
      </>
    );
  }, [data.name, data.salesHistorySummaries]);

  return (
    <section className={styles.container}>
      <h3>
        <SidelineHeading>間取り・広さ</SidelineHeading>
      </h3>
      {content}
    </section>
  );
};

export default UnitSize;
