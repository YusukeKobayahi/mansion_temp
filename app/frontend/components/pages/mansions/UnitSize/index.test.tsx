import { render } from "@testing-library/react";
import mockData from "../fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import UnitSize from ".";

describe("components/pages/mansions/UnitSize", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(<UnitSize data={mock.mansion} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
