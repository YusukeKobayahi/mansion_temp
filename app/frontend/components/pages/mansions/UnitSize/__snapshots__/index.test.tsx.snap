// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/mansions/UnitSize 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <h3>
      <div
        class="heading"
      >
        <span
          class="inner"
        >
          間取り・広さ
        </span>
      </div>
    </h3>
    <p
      class="info"
    >
      間取りには1K～1DKがあり、一人暮らしを検討している方～部屋を活用する一人暮らし・二人暮らしを検討している方におすすめのマンションです。
      <br />
      広さでみると20.72㎡～45.16㎡の物件があり、リヴシティ神田では表内の色がついている広さの部屋が存在しています。
    </p>
    <table
      class="table"
    >
      <tbody>
        <tr>
          <th
            class="hasUnit"
          >
            ～39㎡
          </th>
          <td>
            一人暮らしに向いており、特に30㎡以上だとゆったり使えるスペースがあります。
          </td>
        </tr>
        <tr>
          <th
            class="hasUnit"
          >
            40～59㎡
          </th>
          <td>
            二人暮らしに向いている広さです。リビング・ダイニングの他に1,2部屋つく目安です。
          </td>
        </tr>
        <tr>
          <th
            class=""
          >
            60～89㎡
          </th>
          <td>
            家族3人でも過ごしやすく、お子様がいる場合にもおすすめです。
          </td>
        </tr>
        <tr>
          <th
            class=""
          >
            90㎡～
          </th>
          <td>
            家族4人以上で暮らす場合にも適しており、色々な用途でスペースを使い分けられます。
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;
